### PR TITLE
if all pre-req caps granted or pending - then top cap pending

### PR DESCRIPTION
**Pull request by Nauna**
Consider Canadian Onboarding,
When Onboarding pre-req capabilities all either granted or pending, this changes the Onboarding capability to pending.

### CODE REVIEW

**Comment by noodlemoodle** [comment]
thinking this logic can be merged with checkPrereqs since we are doing pretty similar operations twice.
maybe we can have only one method that return the "lowest" status out of all the prereqs, in an order like [expired, action_required, pending, granted]. so that we can call it once in put_when checking prereqs and save the returned status, then later we can use the saved variable to check if should be granted or pending or action required?

**Comment by Nauna** [comment]
So now instead of `checkPrereqs` we have `checkPrereqsChainedStatus`

So now checkPrereqsChainedStatus returns one of 3 status.
1) GRANTED: if all pre-reqs are in granted status
2) PENDING: if all pre-reqs are either granted or pending status
3) ACTION_REQUIRED: if not any of the above

File: src/foam/nanos/crunch/UserCapabilityJunctionDAO.js [hunk 1]
```diff
@@ -311,40 +317,42 @@ foam.CLASS({
           type: 'foam.core.FObject'
         }
       ],
-      type: 'Boolean',
+      type: 'CapabilityJunctionStatus',
       documentation: `Check if prerequisites of a capability is fulfilled`,
```
Comment (noodlemoodle) [inline src/foam/nanos/crunch/UserCapabilityJunctionDAO.js]: please update

**Comment by noodlemoodle** [comment]
haven't tested yet, but logic of it looks good to me

File: src/foam/nanos/crunch/UserCapabilityJunctionDAO.js (hunk 1 above)
Comment (Nauna) [reply src/foam/nanos/crunch/UserCapabilityJunctionDAO.js]: done - thx for this!

**Comment by jlhughes** [comment]
> So now instead of `checkPrereqs` we have `checkPrereqsChainedStatus`
> 
> So now checkPrereqsChainedStatus returns one of 3 status.
> 
> 1. GRANTED: if all pre-reqs are in granted status
> 2. PENDING: if all pre-reqs are either granted or pending status
> 3. ACTION_REQUIRED: if not any of the above

so is 2 - granted and at least one pending?

File: src/foam/nanos/crunch/UserCapabilityJunctionDAO.js [hunk 2]
```diff
@@ -311,40 +317,47 @@ foam.CLASS({
           type: 'foam.core.FObject'
         }
       ],
-      type: 'Boolean',
-      documentation: `Check if prerequisites of a capability is fulfilled`,
+      type: 'CapabilityJunctionStatus',
+      documentation: `
+        Check statuses of all preRequist capabilities - returning:
+        GRANTED: if all pre-reqs are in granted status
+        PENDING: if all pre-reqs are either granted or pending status
```
Comment (jlhughes) [inline src/foam/nanos/crunch/UserCapabilityJunctionDAO.js]: At least one pre-req is still pending.

**Review by jlhughes — APPROVED** [review]

File: src/foam/nanos/crunch/CapabilityCategory.js [hunk 3]
```diff
@@ -75,7 +75,6 @@ foam.CLASS({
       javaThrows: ['AuthorizationException'],
       javaCode: `
         if ( getVisibilityCondition() == null ) {
-          ( getDefaultAuthorizer()).authorizeOnRead(x, this);
           return;
```
Comment (Nauna) [inline src/foam/nanos/crunch/CapabilityCategory.js:78]: ### Just a note on this:
Either we added **capabilitycategory.read.*** to each group that needed to see crunch categories - or we accept that everyone should see categories.

Latter makes more sense - discussed originally by @noodlemoodle - just appending this to this pr

**Pull request merged**